python/TestHarness/schedulers/Scheduler.py

```diff
@@ -227,74 +352,120 @@ def doTimeoutTests(self, tester):
+        # Test is Finished (failed, passed, skipped, silent, etc)
         if tester.isFinished():
-            # And the running caveat and
+            # And the running caveat
             if tester_name in self.reported_jobs and tester.didPass():
                 tester.specs.addParam('caveats', ['FINISHED'], "")
 
             # Print finalized results
-            # TODO: this prints TIMEOUT tests twice. Fix that.
             self.harness.handleTestStatus(tester)
 
-        else:
-            # place this job back in the runner/status queue because it was not ready to launch or is pending
+            # Remove this finished tester from the DAG and active job set, get the next node, and assign it to
+            # a runner thread
             with self.thread_lock:
-                # Make sure the job isn't already in the runner queue (job that is not skipped)
-                if tester.getTestName() not in self.active_jobs:
-                    self.runnerGo(tester_name)
-
-                # Place this job back in the status queue (job is running and is not done)
-                else:
-                    self.statusGo(tester_name)
+                tester_dag = self.shared_testers_dag[tester.getTestDir()]
+
+                # Handle dependencies that will fail if this test did not successfully finish but only if we care
+                # about prereqs
+                if not tester.didPass():
+                    self.deleteDownstreamTests(tester_dag, tester)
+
+                # Get the next list of runnable nodes (jobs)
+                tester_dag.delete_node(tester_name)
+                next_jobs = tester_dag.ind_nodes()
+
+                # Loop through and determine if we need to assign a thread to this new job
+                new_tmp_jobs = []
+                for job in next_jobs:
+                    tmp_tester = self.scheduled_name_to_object[job]
+                    if job not in self.active_jobs and tmp_tester.isInitialized():
+                        tmp_tester.setStatus('QUEUED', tmp_tester.bucket_pending)
+                        new_tmp_jobs.append(job)
+
+            # Add new jobs outside thread lock
+            for job in new_tmp_jobs:
+                self.runnerGo(job)
+
+        # Test is Pending and could not run for some reason or another
+        else:
+            # The runner when performing checkRunnableBase, didn't fail or skip the test. But it couldn't
+            # run for other reasons (not enough slots, etc). So place it back in the runner queue
+            if tester_name not in self.active_jobs:
+                self.runnerGo(tester_name)
 
             # We should spin here for just a bit; A runner and status thread will rapidly hand off a
-            # job to each other for tests that can't run yet due to a prereq not being ready.
-            sleep(0.01)
+            # job to each other for tests that can't run yet due to not enough slots being available
+            #sleep(0.01)
```

Comment:
I was testing a no sleep. @permcody Its as you thought... I don't think we need it. I had a print statement here that never really fired. Which is awesome! That means hardly any testers were pending.